src/os/layer/vectortile.js

```diff
@@ -0,0 +1,727 @@
+goog.module('os.layer.VectorTile');
+goog.module.declareLegacyNamespace();
+
+goog.require('os.ui.layer.defaultLayerUIDirective');
+
+const GoogEventType = goog.require('goog.events.EventType');
+const {getRandomString} = goog.require('goog.string');
+
+const {DEFAULT_MAX_ZOOM, DEFAULT_MIN_ZOOM} = goog.require('ol');
+const events = goog.require('ol.events');
+const olExtent = goog.require('ol.extent');
+const Property = goog.require('ol.layer.Property');
+const VectorTileLayer = goog.require('ol.layer.VectorTile');
+const VectorTileLayerRenderer = goog.require('ol.renderer.canvas.VectorTileLayer');
+const UrlTileSource = goog.require('ol.source.UrlTile');
+
+const {dispatcher} = goog.require('os');
+const IGroupable = goog.require('os.IGroupable');
+const ActionEventType = goog.require('os.action.EventType');
+const DataManager = goog.require('os.data.DataManager');
+const LayerEvent = goog.require('os.events.LayerEvent');
+const LayerEventType = goog.require('os.events.LayerEventType');
+const PropertyChangeEvent = goog.require('os.events.PropertyChangeEvent');
+const {reduceExtentFromLayers} = goog.require('os.fn');
+const osImplements = goog.require('os.implements');
+const {SynchronizerType, identifyLayer} = goog.require('os.layer');
+const ExplicitLayerType = goog.require('os.layer.ExplicitLayerType');
+const ILayer = goog.require('os.layer.ILayer');
+const LayerType = goog.require('os.layer.LayerType');
+const LayerPropertyChange = goog.require('os.layer.PropertyChange');
+const osMap = goog.require('os.map');
+const math = goog.require('os.math');
+const registerClass = goog.require('os.registerClass');
+const SourcePropertyChange = goog.require('os.source.PropertyChange');
+const {isStateFile} = goog.require('os.state');
+const TimeInstant = goog.require('os.time.TimeInstant');
+const ui = goog.require('os.ui');
+const Icons = goog.require('os.ui.Icons');
+const renamelayer = goog.require('os.ui.renamelayer');
+
+
+/**
+ * OpenSphere vector tile layer.
+ *
+ * @implements {ILayer}
+ * @implements {IGroupable}
+ */
+class VectorTile extends VectorTileLayer {
```

Comment:
This isn't something we can fix here, but the amount of boilerplate code we've copied around for our tile layer classes makes me sad. This is 8 or 9 of them now that have identical interface code.

Reply:
Yeah I'm not a fan either. We would have to add interface methods to the lower level OL classes to fix it, which would mean more mixins. Or the code would have to be moved off of the layer (not sure where). Will had originally reused methods from `os.layer.Tile` by assigning them off the prototype, but that felt a little wonky too and I had to change some of the implementations to suit vector tiles.

Reply:
There's a fairly nice way to do this in JS but the Closure Compiler doesn't like it.